presto-product-tests/src/main/java/com/facebook/presto/tests/SqlCancelTests.java

```diff
@@ -0,0 +1,197 @@
+        if (queryCanceller != null) {
+            queryCanceller.close();
+        }
+        query(format("DROP TABLE IF EXISTS %s", TABLE_NAME));
+    }
+
+    private class SqlExecuteCallable
+            implements Callable<Boolean>
+    {
+        private String sql;
+
+        public SqlExecuteCallable(String sql)
+        {
+            this.sql = sql;
+        }
+
+        public Boolean call()
+        {
+            try {
+                query(sql);
+                return false;
+            }
+            catch (QueryExecutionException e) {
+                return true;
+            }
+        }
+    }
+
+    private class SqlCancelCallable
+            implements Callable<Response>
+    {
+        private String sqlStatement;
+
+        public SqlCancelCallable(String sqlStatement)
+        {
+            this.sqlStatement = sqlStatement;
+        }
+
+        public Response call()
+        {
+            long endTime = System.currentTimeMillis() + WAIT_TIME;
+            while (System.currentTimeMillis() < endTime && !Thread.currentThread().isInterrupted()) {
+                List<Object> queryResult = query(format("SELECT state, query_id from system.runtime.queries" +
+                        " WHERE query = '%s' ORDER BY created DESC LIMIT 1", sqlStatement)).row(0);
+                if (queryResult == null || queryResult.size() != 2) {
+                    sleepUninterruptibly(STEP, TimeUnit.MILLISECONDS);
+                    continue;
+                }
+                String queryState = (String) queryResult.get(0);
+                String queryId = (String) queryResult.get(1);
+                if (!queryState.equalsIgnoreCase("running")) {
+                    sleepUninterruptibly(STEP, TimeUnit.MILLISECONDS);
+                }
+                else {
+                    return queryCanceller.cancel(queryId, serverAddress);
+                }
+            }
+            return null;
+        }
+    }
+
+    private void doCancelTest(String sql)
+            throws Exception
+    {
+        Future<Boolean> executeQueryThread = executor.submit(new SqlExecuteCallable(sql));
+        Future<Response> cancelQueryThread = executor.submit(new SqlCancelCallable(sql));
+
+        Boolean isQueryCancelled;
+        try {
+            isQueryCancelled = executeQueryThread.get(2, TimeUnit.MINUTES);
+        }
+        catch (TimeoutException e) {
+            executeQueryThread.cancel(true);
+            throw new RuntimeException("Query execution time limit exceeded", e);
+        }
+
+        String queryState = (String) query(format("SELECT state from system.runtime.queries" +
+                " WHERE query = '%s' ORDER BY created DESC LIMIT 1", sql)).row(0).get(0);
+        assertEquals(queryState.equalsIgnoreCase("failed"), true, "Failed to cancel query");
+        assertEquals(isQueryCancelled.booleanValue(), true, "Failed to cancel query");
+
+        Response response = null;
+        response = cancelQueryThread.get();
```

Comment:
instead of setting it to null first, just declare
`Response response = cancelQueryThread.get()`